R/run.R

```diff
@@ -20,16 +20,68 @@ run.default <- function(x, ...) {
 }
 
 
+#' @param publish Flag for whether to publish results to a Conbench server. See
+#' "Environment Variables" section for how to specify server details. Requires
+#' the benchconnect CLI is installed; see [install_benchconnect()].
+#' @param run_name Name for the run. If not specified, will use `{run_reason}: {commit hash}`
+#' @param run_reason Required. Low-cardinality reason for the run, e.g. "commit" or "test"
+#'
+#' @section Environment Variables:
+#'
+#' - `CONBENCH_URL`: Required. The URL of the Conbench server with no trailing
+#' slash. For arrow, should be `https://conbench.ursa.dev`.
+#' - `CONBENCH_EMAIL`: The email to use for Conbench login. Only required if the
+#' server is private.
+#' - `CONBENCH_PASSWORD`: The password to use for Conbench login. Only required
+#' if the server is private.
+#' - `CONBENCH_PROJECT_REPOSITORY`: The repository name (in the format
+#' `org/repo`) or the URL (in the format `https://github.com/org/repo`).
+#' Defaults to `"https://github.com/apache/arrow"` if unset.
+#' - `CONBENCH_PROJECT_PR_NUMBER`: Recommended. The number of the GitHub pull
+#' request that is running this benchmark, or `NULL` if it's a run on the
+#' default branch
+#' - `CONBENCH_PROJECT_COMMIT`: The 40-character commit SHA of the repo being
+#' benchmarked. If missing, will attempt to obtain it from
+#' `arrow::arrow_info()$build_info$git_id`, though this may not be populated
+#' depending on how Arrow was built.
+#' - `CONBENCH_MACHINE_INFO_NAME`: Will override detected machine host name sent
+#' in `machine_info.name` when posting runs and results. Needed for cases where
+#' the actual host name can vary, like CI and cloud runners.
+#'
 #' @rdname run
 #' @export
-run.BenchmarkDataFrame <- function(x, ...) {
+run.BenchmarkDataFrame <- function(x, ..., publish = FALSE, run_name = NULL, run_reason = NULL) {
   # if already run (so no elements of `parameters` are NULL), is no-op
   x <- get_default_parameters(x, ...)
 
+  if (publish) {
+    github <- github_info()
+    if (is.null(run_name)) {
+      run_name <- paste(run_reason, github$commit, sep = ": ")
+    }
+    stopifnot(
+      "Results cannot be published without a `run_reason`!" = !is.null(run_reason)
+    )
+    bm_run <- BenchmarkRun$new(name = run_name, reason = run_reason, github = github)
+    start_run(run = bm_run)
+  }
+
   x$results <- purrr::map2(x$benchmark, x$parameters, function(bm, params) {
-    run_benchmark(bm = bm, params = params, ...)
+    withr::with_envvar(
+      c(CONBENCH_RUN_NAME = run_name, CONBENCH_RUN_REASON = run_reason),
+      { ress <- run_benchmark(bm = bm, params = params, ...) }
+    )
+    if (publish) {
+      ress$results <- lapply(ress$results, augment_result)
+      lapply(ress$results, submit_result)
+    }
+    ress
```

Comment:
Is it possible to pass run_name and run_reason through in the results JSON? It's not a huge deal, but it seems a little bit funny that we would pass information through both env vars and JSON here rather then relying on an all-json approach (or in principle an all env-var approach, but that would be silly!).

I know that there are good reasons to have things like `CONBENCH_RUN_NAME` exist independently, but I thought that `benchconnect` allowed both specifying through json or it would pick up the defaults form the envvars if nothing was passed through the json. Am I confused about that?

Reply:
Also: what's the relation with the env vars being used in `run` below? If I'm reading this correctly it looks like those use the unprocessed versions of those env vars (e.g. before lines 58-63 process some of the contents of them)?

Reply:
> Is it possible to pass run_name and run_reason through in the results JSON? It's not a huge deal, but it seems a little bit funny that we would pass information through both env vars and JSON here rather then relying on an all-json approach (or in principle an all env-var approach, but that would be silly!).
> 
> I know that there are good reasons to have things like `CONBENCH_RUN_NAME` exist independently, but I thought that `benchconnect` allowed both specifying through json or it would pick up the defaults form the envvars if nothing was passed through the json. Am I confused about that?

Yeah, this was one of the weird design decisions I made. So in theory, it is possible to pass these all the way through as arguments to each level of the runner. But at the bottom, we actually need to set them as env vars anyway to make sure they get passed across to the new process created by `run_script()` so they'll get picked up in the `BenchmarkResult$new()` call in the `run_bm()` call in the script. And we need them in-between, too, because there's a second `BenchmarkResult$new()` call in the original process for when the script errors.

There are other potential ways besides env vars to pass them across to `run_bm()`, e.g. some form of currying, but they mostly seem worse. And if we take that decision as fixed, then changing all the intermediate signatures to just pass them through seems superfluous when we can just let the env vars carry them through. But this is an odd design, so if we'd rather, I can rearrange all the signatures and create these env vars in `run_script()`.

> Also: what's the relation with the env vars being used in `run` below? If I'm reading this correctly it looks like those use the unprocessed versions of those env vars (e.g. before lines 58-63 process some of the contents of them)?

I don't think I understand the question? These mostly don't change, although when we consume them we do change `""` (which `Sys.getenv()` will return even if it's not set) to `NULL`.

Reply:
Oh! Yeah, that's because I started using `f"{run_reason}: {github$commit}"` as a default `run_name` on the Python side for legacy reasons and, well, anything is better than nothing, so I ported it over here so we're not doing the same thing in arrow-benchmarks-ci or wherever just so there's _something_ in there. 

We could stick that default in `BenchmarkRun` and `BenchmarkResult` if we want, though that does get a bit tricky about handling updating (e.g. if `run_reason` is updated, `run_name` probably should get updated if it was a previously-generated default, but not if it was previously specified, so you have to be able to tell the difference). That felt overly complicated, so, well, this. 

Another possibility would be to do it in the signature with lazy eval like

```r
run.BenchmarkDataFrame <- function(x, 
                                   ..., 
                                   publish = FALSE, 
                                   github = github_info(), 
                                   run_reason = NULL, 
                                   run_name = paste(run_reason, github$commit, sep = ": ")) {
```

but the drawback there is `github$commit` will be `NULL` for any non-source Arrow installation because `arrow::arrow_info()$build_info$git_id` isn't populated on precompiled binaries. Adding that users running but not publishing will not bother to set `run_reason` either, and their default `run_name` will be `": "`, which is pretty silly. We could do 

```r
                                   run_name = if (publish) paste(run_reason, github$commit, sep = ": ")) {
``` 
instead, but maybe that's too clever?

Anyway, sorry that's a bit long-winded, but that's my thought process. If any of the alternatives seem desirable to you, I'm happy to change it; otherwise I'll put a comment in there on why if I can figure out how to make it shorter than this _Treatise on Defaults for `run_name`_.

***

> I'm fine to leave it as is if it's the least-bad option (especially as we transition to this newer way of running things). It might be helpful to have some of the context you put here in comment(s) explaining why what looks a bit off is the best option for now.

Honestly I'm not totally convinced myself that I made the right call here; maybe I should just change all the signatures. I might try and see if I find that more distasteful than this, and if so, put an explanation in a comment here.

Another option here would be to force these params to be passed via env var (which is a bit extreme), or allow it by making defaults `run_name = Sys.getenv("CONBENCH_RUN_NAME")`, run_reason = Sys.getenv("CONBENCH_RUN_REASON")`. Either way, it would be promoting them to first class env vars exposed to external users instead of just used internally, and would make the arrow-benchmarks-ci a little more concise, basically 

```r
bm_df <- get_package_benchmarks()
bm_df <- bm_df[bm_df$name %in% c("vector", "of", "benchmark", "names"), ]
run(bm_df, publish = TRUE)
```
instead of explicitly splicing in all the env vars, at the cost of making it even harder to see how all the pieces therein get their metadata—all the github stuff is already hidden this way, as are the Conbench client setup values (URL, user, password). I guess I'm conflicted because regardless it feels like I'm adding to the complexity of this codebase and making it harder to maintain. I'm starting to get the appeal of config files and objects corresponding to them so at least everything is in one place. (Not that I'm suggesting that here; it's very un-R-ic.)

Reply:
Hot take. I know you've put lots of effort/code into ensuring run name has sensible defaults, but because of a lot of recent work we really don't need `run_name` to be "nice" anymore. In fact, it's almost superfluous. It doesn't change any Conbench behavior anymore, and it's only shown in the webapp on the Benchmark page, where you can also easily see the run reason and commit hash if you want to.

So if it would be easier/cleaner to soft-deprecate run name, leave it blank or make it a UUID, and not make it possible for a user to change that, maybe we should consider that.

Reply:
I'm cool with nulling it out, but code like [this](https://github.com/conbench/conbench/blob/main/conbench/entities/run.py#L274) makes me nervous it will cause issues somewhere. Making it a UUID seems duplicative when we've already got `run_id`.

Reply:
Oh good catch, there is one place it's used -- the bottom of the `/compare/runs/A...B` page in a picker widget.

I just made https://github.com/conbench/conbench/issues/607 to look more into deprecation of run_name on the conbench side. In the meantime, I still think it already is redundant and everything would be fine with a UUID here.

Reply:
Without seeing the actual code-consequences here I'm finding it a bit hard to see the advantages of each of the options listed. It sounds like (though correct me if I'm wrong, there were a number of options proposed and I might not be following the consequences of each exactly perfectly!) we have a few less-than-good options where the complexity is either in passed signatures or using env vars for some things but not other things and have that funky bifurcation.

> Another option here would be to force these params to be passed via env var (which is a bit extreme), or allow it by making defaults run_name = Sys.getenv("CONBENCH_RUN_NAME"), run_reason = Sys.getenv("CONBENCH_RUN_REASON")`. Either way, it would be promoting them to first class env vars exposed to external users instead of just used internally, and would make the arrow-benchmarks-ci a little more concise, basically

This (use envvar, making the defaults bit) seems least bad so far (especially for what we need to accomplish right now). Currently the only callers of arrowbench that need to publish are ci systems (and specifically only arrow-benchmarks-ci) so I'm not particularly worried about that interface being annoying to use by humans (yet). 

I would be in favor of doing the least-bad thing here (commenting heavily where there is even a little bit of funk) and getting this stuff merged in + actually running (since currently all of this is theoretical and never being called anywhere). I'm certain we will learn as we use this for real, and we can revisit + polish this design when we get real feedback.

Reply:
Alright, managed to rip out the new env vars wholesale by adding to signatures. It's not horrible, and I think is less confusing.